core/engine/src/vm/opcode/args.rs

```diff
@@ -807,3 +939,654 @@ impl Argument for (VaryingOperand, ThinVec<u32>) {
         ((first.into(), rest), pos + 6 + total_len * 4)
     }
 }
+
+#[cfg(test)]
+mod test {
+
+  use std::vec;
+  use thin_vec::thin_vec;
+  use super::*;
+
+  fn test_write_arg<T: Argument>(test_data: T) -> Vec<u8> {
+    let mut v: Vec<u8> = Vec::new();
+    test_data.encode(&mut v);
+    return v
+  }
+
+  // This can fail depending on Argument/opcode combination given (too short)
+  fn test_read_arg<T: Argument>(opcode: Box<[u8]>, pos: usize) -> (T, usize) {
+    let r: &[u8] = &opcode;
+    T::decode(r, pos)
+  }
+
+
+  //
+  // VariableOperand
+  //
+  #[test]
+  fn arg_vop_u8() {
+    test_write_arg(VaryingOperand{value: u8::MAX as u32});
+    test_read_arg::<VaryingOperand>(Box::new([0,0xFF]), 0);
+  }
+
+  #[test]
+  #[should_panic]
+  fn fail_arg_vop_u8() {
+    test_read_arg::<VaryingOperand>(Box::new([0]), 0);
+  }
+
+  #[test]
+  fn arg_vop_u16() {
+    test_write_arg(VaryingOperand{value: u16::MAX as u32});
+    test_read_arg::<VaryingOperand>(Box::new([1,0xFF,0xFF]), 0);
```

Comment:
We should check if the argument is equal after turning to bytes and back.
```suggestion
    let arg = VaryingOperand::from(u16::MAX);
    test_write_arg(arg);
    assert_eq(test_read_arg::<VaryingOperand>(Box::new([1,0xFF,0xFF]), 0), arg);
```